src/clr-angular/utils/focus/key-focus/key-focus.ts

```diff
@@ -0,0 +1,176 @@
+      this._focusChange.next();
+    }
+  }
+
+  ngAfterContentChecked() {
+    this.getChildren().forEach(button => {
+      this.setTabIndex(button, -1);
+    });
+    this.setTabIndex(this.getCurrent(), 0);
+  }
+
+  @HostListener('keydown', ['$event'])
+  handleKeyboardEvent(event: any) {
+    if (this.prevClicked(event) && this._current - 1 >= 0) {
+      // go to previous unless at first item
+      this.keyAction(() => {
+        this._current--;
+      });
+    } else if (this.nextClicked(event) && this._current + 1 < this.getChildren().length) {
+      // go to next unless at last item
+      this.keyAction(() => {
+        this._current++;
+      });
+    } else if (event.key === 'Home') {
+      this.keyAction(() => {
+        this._current = 0;
+      });
+    } else if (event.key === 'End') {
+      this.keyAction(() => {
+        this._current = this.getChildren().length - 1;
+      });
+    }
+  }
+
+  @HostListener('click', ['$event'])
+  handleClickEvent(event: any) {
+    const position = this.getPositionOf(event.srcElement as HTMLElement);
+    if (position > -1) {
+      this._current = position;
+    }
+  }
+
+  getPositionOf(element: HTMLElement) {
+    const children = Array.from(this.getChildren());
+    return children.indexOf(element);
+  }
+
+  moveTo(position: number, force: boolean = false) {
+    if (
+      isPlatformBrowser(this.platformId) &&
+      position >= 0 &&
+      position < this.getChildren().length &&
+      (force || position !== this._current)
+    ) {
+      this._current = position;
+      this.getCurrent().focus();
+      this._focusChange.next();
+    }
+  }
+
+  private keyAction(action: Function) {
+    if (!isPlatformBrowser(this.platformId)) {
+      return;
+    }
+    this.setTabIndex(this.getCurrent(), -1);
+    action.call(this);
+    this.setTabIndex(this.getCurrent(), 0);
+    this.getCurrent().focus();
+    this._focusChange.next();
+    // Prevent window scroll on arrow key pressed
+    event.preventDefault();
+    event.stopImmediatePropagation();
+  }
+
+  private prevClicked(event: any) {
+    switch (this.direction) {
+      case ClrFocusDirection.VERTICAL:
+        return event.key === 'ArrowUp';
+      case ClrFocusDirection.HORIZONTAL:
+        return event.key === 'ArrowLeft';
+      case ClrFocusDirection.BOTH:
+        return event.key === 'ArrowUp' || event.key === 'ArrowLeft';
+      default:
+        return false;
+    }
+  }
+
+  private nextClicked(event: any) {
+    switch (this.direction) {
+      case ClrFocusDirection.VERTICAL:
+        return event.key === 'ArrowDown';
+      case ClrFocusDirection.HORIZONTAL:
+        return event.key === 'ArrowRight';
+      case ClrFocusDirection.BOTH:
+        return event.key === 'ArrowDown' || event.key === 'ArrowRight';
+      default:
+        return false;
+    }
+  }
+
+  private getChildren(): HTMLElement[] {
+    return this.elementRef.nativeElement.querySelectorAll(this.selector) || [];
```

Comment:
I guess my previous explanation disappeared in the resolved comments, so I'll do it again.

About 1st: 
- ViewChildren and ContentChildren are not suitable because:
  - They require us to change API and force the user to start applying directives to their items with the sole purpose of helping us select children in a way that we feel comfortable with. I strongly feel that we should avoid this, especially on already existing components.
  - They operate on the View level. The View level is appropriate for performing structural DOM operations - such as applying TemplateRefs to ViewContainerRefs, creating, moving children, etc.
  - We actually operate on the direct Rendering level. We don't need to go through the View level, just to obtain our nativeElements, it's overkill.
  - Bonus fact is that even Renderer2 has no handlers for focus operations. Ergo => Focus operations need to be pure direct DOM manipulations by definition.
- Traversing trees - this directive is not aimed at traversing a whole tree at once. It should be applied flatly at each specific level. The expand/collapse navigation is not focus-specific operation, so it has to be handled by the tree component itself. This directive will handle only the current active node children at a time. This greatly reduces the expected number of items traversed in a tree.
- Direct DOM access
   - What makes you say that direct DOM access is expensive? Can you point to some official references and researches?
   - Nevertheless, let's put it to a test. A simple stackblitz test I did shows that:
      - in the 10,000-15,000 items the delay starts to become noticeable, but still acceptable.
      - up to 50,000 items the delays are visible and annoying, but research showed that they come from the roving tabindex pattern implementation. We may try to improve on this. Rendering time of the list itself is quite prominent here. Not sure this is a scenario where focus issues are the biggest problem of the user that created such a list.
      - above 50,000 items the mere rendering time of the list is unacceptable. In the 100s of 1000s of items range stackblitz simply dies.

About 2nd
- This directive is not published on the website deliberately. It is intended to be used as internal API. We well know that people (like H5) sometimes use such tools, but this is on their own account.
- It is flexible enough to accept custom selectors.
- Currently we don't have a scenario where menu items will be comprised of nested menus. I hope we never will, this looks like a monstrosity.
- Display: none - I agree the current directive implementation does not handle this. But:
   - As this is internal tool, we can update it where such need arises. It will be incremental update, so it will be safe.
   - Latest trends are that disabled items are selectable (see Jeeyun's dropdown PR)

About 3rd
- The directive gets updated on AfterContentChecked. So it's not deaf to these changes at all.
- A just concern may arise that if ContentChecked is ran way too often, then this directive will cause trouble. I am open to discussing cases where it's justified to have a looped ContentChecked events, which is not a symptom of serious issue itself.

Thanks for the raised concerns, this is complex issue, so I am glad to have many eyes go over it!
We can have a live discussion about it too, if needed.

 

About 2nd:

Reply:
Again to the display: none -> I first got confused that you were talking about "disabled"
But I don't think that we handle the display:none case anywhere - at least with the dropdowns we don't.
We also recommend using disabled over invisible when it comes to temporarily disabling items, so I'm not sure how much we have to handle the invisible case in our focus managements. May bring this to discussion with Chris and the team.

Reply:
> What makes you say that direct DOM access is expensive? Can you point to some official references and researches?
Accessing a large list of DOM elements is simply costly. This is actually the problem that not only Angular but many other UI frameworks try to tackle. http://www.phpied.com/dom-access-optimization/ Angular solves it with view references and change detection. React has its VirtualDOM that keeps dom references in objects.
This is only one of the reasons why Angular preaches “DON’T TOUCH DOM directly”. Angular keeps references to their view elements once and we access them via ElementRef, and Angular Change Detection let us(developers) know if there are any changes in the view. These are all there to prevent devs from accessing DOM’s native elements directly. So traversing DOM children directly using querySelector/getElementsByClassName, again and again, is discouraged, especially there are many of them.
But what we can do is access the native elements through ElementRefs. We do that in many places actually. The reason is they are references already created by Angular **once**. We can now access the native elements via those ElementRefs; we ourselves don’t have to traverse\access them.
Now in our case, if we place this directive on component or element that could nest a large number of items, this has to go down traverse the children on each keydown. Imagine a user keeps pressing down or up to get on the right item. On each single keydown, this directive will do its traversal and change attribute values on them.

Reply:
You also mentioned the use of `AfterContentChecked`. I didn't even see that. That means it will do traversal on every change detection cycle now even if the change detections happens somewhere else as ngDoCheck, AfterContentChecked, AfterViewChecked. runs on each change detection cycle.

Reply:
Forgot to update details about the last commit.

- `AfterContentChecked` was removed in latest update. It was not strictly necessary and I agree it may not be as predictable as we would like it to be. Thanks for that!
- There was also a huge improvement in the roving tabindex implementation, which leads to great decrease in DOM retrieval calls. Thanks for bringing the attention to this!
- We should disambiguate the terms "traversal" vs "retrieval" though. `querySelectorAll` is a bulk request, same as `select from table` in SQL is. It retrieves the required nodes only at one call. While "traversal implies something else. Actually the link you provided "preaches" exactly this - we should not do multiple calls in a JS layer loop:
```
// bad
for (var count = 0; count < 15000; count++) {  
    document.getElementById('here').innerHTML += 'a';  
}
```
Which we do not do.
- Also “DON’T TOUCH DOM directly”, for the rendering case (vs. the layout case, as described in my previous post), means "Please, use Renderer2 if possible". Sadly, Renderer2 does not handle focus management. A workaround would be to implement our own custom renderers, just for this case, and support them for every single platform we are interested in.
- I already mentioned that I made competitive performance tests for both cases - native vs contentchildren and up to 50,000 of items the deviations are in the range of the statistical errors. It's good for it to remain in the task record. Here is the setup:
https://stackblitz.com/edit/queryselector-vs-contentchildren?file=src/app/app.component.ts
- @coryrylan attempted a non-API-breaking alternative, which was promising at first, but was not fully successful because in Tabs the buttons are not direct children of the component and the portal case (nested tabs) gets broken with `descendants: true`. I remembered that I'd researched this case in the beginning of this task and gone through the same path. Maybe it would be nice to document such researches better when it's about more complex tasks that take some time to make it through the review process.

Best regards 🌴